logdna/logdna.py

```diff
@@ -99,72 +106,61 @@ def buffer_log(self, message):
                 self.internalLogger.debug('Error in calling buffer_log: %s', e)
 
     def buffer_log_sync(self, message):
-        # Attempt to acquire lock to write to buf
-        # otherwise write to secondary as flush occurs
-        if self.lock.acquire(blocking=False):
-            msglen = len(message['line'])
-            if self.buf_size + msglen < self.buf_retention_limit:
-                self.buf.append(message)
-                self.buf_size += msglen
-            else:
-                self.internalLogger.debug(
-                    'The buffer size exceeded the limit: %s',
-                    self.buf_retention_limit)
-
-            if self.buf_size >= self.flush_limit and not self.exception_flag:
-                self.close_flusher()
-                self.flush()
-            else:
-                self.start_flusher()
-            self.lock.release()
-        else:
-            self.secondary.append(message)
+        # Attempt to acquire lock to write to buffer
+        if self.lock.acquire(blocking=True):
+            try:
+                msglen = len(message['line'])
+                if self.buf_size + msglen < self.buf_retention_limit:
+                    self.buf.append(message)
+                    self.buf_size += msglen
+                else:
+                    self.internalLogger.debug(
+                        'The buffer size exceeded the limit: %s',
+                        self.buf_retention_limit)
 
-    def clean_after_success(self):
-        self.close_flusher()
-        self.buf.clear()
-        self.buf_size = 0
-        self.exception_flag = False
+                if self.buf_size >= self.flush_limit:
+                    self.flush()
```

Comment:
I _think_ this is still a small issue, in that this call to `flush()` happens with the lock held, and if `schedule_flush_sync()` is unable to submit the flush to the request thread pool, then it will call `try_lock_and_do_flush_request` itself, which will fail to do anything because the lock is already held.  In theory `submit()` could throw an exception if the pool is shutting down or if it failed to create a new worker thread.  I'm not sure either of those conditions is really worth worrying about, and if the failure mode is just "we don't auto-flush the buffer in that case" that's probably fine -- it's just that the code is a little confused around that case.

I don't feel very strongly about it but if you wanted to address this, you might do something like check the buffer size inside the lock and set a local flag `doFlush = True` as needed; then check that flag and call `flush()` if necessary after releasing the lock.  That would introduce a possible race where multiple flush requests could get queued but that should be harmless as the ultimate flush logic will do nothing if there's nothing to flush.

As I said, I don't feel very strongly about it.  I'm mostly just noting it for posterity I guess.

Reply:
>if the failure mode is just "we don't auto-flush the buffer in that case" that's probably fine -- it's just that the code is a little confused around that case.

Yup. The next buffer flush attempt would occur when either the flush timer thread wakes up, or `buffer_log_message()` is invoked, whichever is handled first. The current default configuration value for this thread is `250ms`, but that can be overridden by environment variables. 

Regarding the potential for attempts to submit flush requests to the thread pool failing, the [docs for python's concurrent futures module](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Executor.shutdown) only state that `RuntimeError` will be raised if `submit()` is invoked after `shutdown()`. The code order in the handler's `close()` method guards against this by ensuring that the executor that is used to handle flush requests can only be shutdown after the executor that handles `buffer_log_message()` requests.  Outside of this, there could be a scenario where there is a system or OS resource issue / limit, but other than that, I think the calling code should always be able to reliably submit the request.  Whether the request succeeds or not will be a different matter.